Make colorbars respect template configurations

joelostblom opened the pull request:
Currently, none of the colorbar options in the templates are effective and colorbars are rendered the same regardless of the template. This PR adds the template colorbar options under the `layout.coloraxis` key instead of assigning them to individual figures. As a bonus the json template files are now about half the size.

I removed the colorbar helper function because I didn't see any other use for it. I also removed the figure level colorscale assignments because it seems like the figures pick up these from the template just fine.

You can see the effect of the proposed changes below.

## ========== Current =================== This PR =======
![image](https://user-images.githubusercontent.com/4560057/68154668-4dd45500-ff48-11e9-9b40-983c09091675.png)

nicolaskruchten commented:
Thanks for the PR!

> Currently, none of the colorbar options in the templates are effective and colorbars are rendered the same regardless of the template

Is this true of *all* colorbars or just the `coloraxis` colorbars? I.e. if you have a colorbar that's defined in a trace, do the current settings work?

joelostblom commented:
> if you have a colorbar that's defined in a trace, do the current settings work?

Hmm, not sure. Could you give me a code test case to run? I saw that `go.Colorbar` is deprecated so not sure what function to use to create a colorbar as a trace rather than in the coloraxis.

nicolaskruchten commented:
All the non-PX code here uses "trace-defined" colorbars rather than `coloraxis` https://plot.ly/python/colorscales/

joelostblom commented:
Thanks!

I tried the "Colorscale for Scatter Plots" example with `colorbar=dict(title="Colorbar"). it seems to be are unaffected by the changes I made and ignore templates both before and after this PR:

Before:
![image](https://user-images.githubusercontent.com/4560057/68161478-4320bc80-ff56-11e9-9959-e6ae5ca58886.png)

After:
![image](https://user-images.githubusercontent.com/4560057/68161899-1d47e780-ff57-11e9-9c38-6dbdd39a2c86.png)

nicolaskruchten commented:
OK so the problem with the example above is that the current builtin templates don't set a colorscale default for `scatter`, only the listed trace types. I don't agree with removing anything from the current templates (this would be a breaking change), although I don't mind adding something to cover `coloraxis.colorbar`

nicolaskruchten commented:
For example, the current settings do work for heatmaps which are not generated with PX:

![image](https://user-images.githubusercontent.com/203523/68173084-c686e180-ff47-11e9-97b7-f7267c50df2c.png)

joelostblom commented:
Ah ok I see. I didn't realize that the colorscale would not work with `go.Heatmap` when it was working with `go.Scatter` and `px.density_heatmap`. Sorry, didn't mean to remove any functionality.

I reverted all removals, rebased, and force pushed. The current version lets the template modify the colorbars and colorscales of both `go.Heatmap` and `px.density_heatmap`.

nicolaskruchten commented:
Can you add the colorbar settings to the other traces types too please? even if only the `scatter*` family? :)

joelostblom commented:
Sorry, I don't understand, can you elaborate on the changes required?

The PR already works with the `px.scatter` colorbar.

![image](https://user-images.githubusercontent.com/4560057/68236734-03191280-0006-11ea-8bf0-73c1a968629d.png)

The colorscale is already picked up by `go.Scatter` so don't think I need to add that separately as for the heatmaps.

![image](https://user-images.githubusercontent.com/4560057/68236891-51c6ac80-0006-11ea-9eeb-a268264f4fc3.png)

nicolaskruchten commented:
Sorry if this isn't clear...

There are two ways to draw a colorbar: "at the trace level" via
something.colorbar attributes on a go.Something (i.e. marker.colorbar in
go.Scatter) or "at the layout level" via layout.coloraxis.colorbar if a
coloraxis is being used.

At the moment, the built-in templates define the colorbar settings (borders
and ticks and so on) at the trace level for certain trace types
(histogram2d, heatmap etc: the ones you had deleted and I had asked you to
undelete). Your PR adds the colorbar settings to the layout level. PX uses
a layout-level colorbar, so this is why PX didn't pick things up before
your PR but does after your PR. This still leaves colorbars that arise from
usage of go.Scatter *without* using PX, i.e. with
`go.Figure(go.Scatter(x=[1,2], y=[1,2], marker_color=[1,2],
marker_showscale=True))` ... in order for such a figure to get the nice
tick labels etc, we'd need to duplicate those lines I had asked you to
undelete, once per trace type that supports a colorbar. Don't worry about
it for now though, this can be added in a future PR.

On Tue, Nov 5, 2019 at 1:57 PM Joel Ostblom <notifications@github.com>
wrote:

> Sorry, I don't understand, can you elaborate on the changes required?
>
> The PR already works with the px.scatter colorbar.
>
> [image: image]
> <https://user-images.githubusercontent.com/4560057/68236734-03191280-0006-11ea-8bf0-73c1a968629d.png>
>
> The colorscale is already picked up by go.Scatter so don't think I need
> to add that separately as for the heatmaps.
>
> [image: image]
> <https://user-images.githubusercontent.com/4560057/68236891-51c6ac80-0006-11ea-9eeb-a268264f4fc3.png>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/plotly.py/pull/1867?email_source=notifications&email_token=AABRWA6YKJMS4WUILUHCHGDQSG6ZZA5CNFSM4JIYXS42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDD5SMA#issuecomment-549968176>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRWA3O24KBJWKXTTU6KCTQSG6ZZANCNFSM4JIYXS4Q>
> .
>


-- 
Nicolas Kruchten - VP Product
e nicolas@plot.ly
5555 Gaspe Ave #118, Montreal, QC, H2T 2A3
<https://github.com/plotly?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://twitter.com/plotlygraphs?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.instagram.com/plotly?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.linkedin.com/company/plotly/?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
